src/02-common-principles.md

```diff
@@ -383,6 +383,200 @@ In particular, if a BIDS dataset contains a `derivatives/` sub-directory,
+an empty string (`""`) MUST NOT be a key in the `DatasetLinks` object.
+
+In the case where a derivatives dataset is nested under a raw dataset and both have a `dataset_description.json` file,
+the BIDS URIs within the nested derivatives dataset MUST be resolved with respect to `/derivatives/dataset_description.json`,
+and the BIDS URIs for the raw dataset (excluding `/derivatives`) MUST be resolved with respect to `/dataset_description.json`.
+
+However if a derivatives dataset is nested under a raw dataset and only the raw dataset has a `dataset_description.json` file,
+all BIDS URIs MUST be resolved with respect to the `/dataset_description.json` file.
+
+(Assuming that in the two above examples `/` refers to the root of the BIDS raw dataset with a nested derivatives dataset.)
+
+### Refer to a file within a dataset
+
+To link for example to a derivative file that is stored in a nested `derivatives/`
+directory within a dataset:
+`bids:deriv1:/sub-01/anat/sub-01_desc-preproc_T1w.nii.gz`
+
+```json
+{
+    "DatasetLinks": {
+        "deriv1": "file://derivatives/derivative1"
+    }
+
+}
+```
+
+Alternatively, we can also make use of the reserved `<dataset-name>`: `""` (an empty string),
+which always points to the root of the current dataset, to point to the same file using the following syntax:
+`bids::/derivatives/derivative1/sub-01/anat/sub-01_desc-preproc_T1w.nii.gz`
+
+Thus, for this specific case there are two ways to refer to the same file:
+
+1.  `bids:deriv1:/sub-01/anat/sub-01_desc-preproc_T1w.nii.gz`,
+    with `deriv1` being specified within `dataset_description.DatasetLinks`,
+    as shown in the example above.
+
+1.  `bids::/derivatives/derivative1/sub-01/anat/sub-01_desc-preproc_T1w.nii.gz`,
+    with `""` (an empty string) NOT being a key in `dataset_description.DatasetLinks`,
+    because it is a reserved value that always points to the root of the current dataset.
+
+**If *all* BIDS URIs in a dataset *only* use an empty string (`""`) as a `<dataset-name>`,**
+**the `DatasetLinks` metadata MAY not be specified.**
+
+### Refer to a remote dataset
+
+To link for example to a derivative file that is stored in a `derivatives/`
+directory that is stored on a remote server called `mydatahost.com`:
+`bids:deriv2:/sub-01/anat/sub-01_desc-preproc_T1w.nii.gz`
+
+```json
+{
+    "DatasetLinks": {
+        "deriv2": "https://mydatahost.com/derivatives/derivative2"
+    }
+}
+```
+
+This means that the needed files can be obtained by downloading the data
+from `https://mydatahost.com/derivatives/derivative2`, and then navigating
+to `/sub-01/anat/sub-01_desc-preproc_T1w.nii.gz` relative to that
+downloaded data.
+
+However,
+arbitrary remote server locations such as the `mydatahost.com` example server may not be stable.
+Therefore it is RECOMMENDED to instead point to archived versions of a datasets using their DOIs:
+
+```json
+{
+    "DatasetLinks": {
+        "deriv3": "doi:10.18112/openneuro.ds003669.v1.0.0"
```

Comment:
Combining the deriv2 and deriv3 examples, the specified example for deriv3 does not resolve. The example [dataset](https://openneuro.org/datasets/ds003669/versions/1.0.0) does not have a `sub-01` folder.

Even if it would have had a `sub-01` folder, it seems to me that it would not resolve. The DOI  resolves as `https://openneuro.org/datasets/ds003669/versions/1.0.0`, but manually looking at the page I can see that the file should be downloaded from `https://openneuro.org/crn/datasets/ds003669/snapshots/1.0.0/files/sub-MP02312:anat:sub-MP02312_T1w.json`. 

There would be two reasons for the download failing: 1) the first section to the path is different (among others it includes `crn` and `files`), 2) the 2nd section to the path contains `:` semicolons rather than slashes, and replacing the `:` with`/` results in `Cannot GET /crn/datasets/ds003669/snapshots/1.0.0/files/sub-MP02312/anat/sub-MP02312_T1w.json`.

I cannot come up with a scheme for this example that would actually work, so to me it seems that the example reveals that this does not work with openneuro.

Reply:
> I cannot come up with a scheme for this example that would actually work, so to me it seems that the example reveals that this does not work with openneuro.

This points to an important distinction between two ways to make BIDS URIs work for **remote locations**.

- One is easy to do and understand for humans, and easy to specify, but not always easily resolvable by tools.
- The other one is harder (maybe sometimes impossible) to do for humans, and (much) harder to specify, but always resolvable by tools.

**Example for 1:**

-    We specify a location on the web (http, https, doi) and expect the user to download from that location.
-    In a second step, the user should use an additionally supplied path to go into their freshly downloaded data.
-    Example location (data to download): https://github.com/mgxd/brainplot/archive/refs/heads/master.zip
-    Example path (absolute with respect to downloaded data): `/brainplot-master/brainplot/Conte69_Atlas/Conte69.R.midthickness.32k_fs_LR.surf.gii`


**Example for 2:**

- We specify a location on the web (http, https, doi) and expect the combination of that location and an additionally supplied path to point to a downloadable file
- This file should be downloadable with tools like curl or wget, or similar.
- Example location: `https://github.com/mgxd/brainplot/raw/master/brainplot`
- Example path: `/Conte69_Atlas/Conte69.R.midthickness.32k_fs_LR.surf.gii`
- Their combination: https://github.com/mgxd/brainplot/raw/master/brainplot/Conte69_Atlas/Conte69.R.midthickness.32k_fs_LR.surf.gii

Note that in both examples you get the exact same file (`Conte69.R.midthickness.32k_fs_LR.surf.gii`).

---

My present proposals implements the "way 1" --> Good human readability, easy specification - at the expense of worse (sometimes impossible) machine readability.

I picked this way (@effigies agreed), because "way 2" (always make BIDS URIs machine resolvable) seemed very hard to specify for an adequately large set of remote hosts

What is your opinion on this? Do you think "way 1" is reasonable? Where did you misunderstand the current proposal (that is, where should we add text to clarify the proposal)?

Reply:
sorry, I realize that I am processing your replies in the wrong order. This also relates to my https://github.com/bids-standard/bids-specification/pull/820#discussion_r685961581

Your description here is very clear. "Way 1" works if the pointer is to a dataset, or another clear single entity package (like `master.zip`). But what if there is no such package (i.e. the link is not a landing page), but directly to a file (e.g. on [S3](https://stackoverflow.com/questions/57224790/how-to-use-an-s3-uri) or on FTP)? 

In `bids:location:path` the example location: `s3://openneuro.org/ds003709/` and path `/sub-23969/ses-visit1/anat/sub-23969_ses-visit1_run-1_T1w.nii.gz`is identical to `s3://openneuro.org/ds003709/sub-23969/ses-visit1/anat` and path `/sub-23969_ses-visit1_run-1_T1w.nii.gz` but also to location `s3://openneuro.org` and path `/ds003709/sub-23969/ses-visit1/anat/sub-23969_ses-visit1_run-1_T1w.nii.gz`. In the latter case and "way 1" it seems I first have to download all(!) of openneuro before I get to interpret the path.  

I prefer that either specification is allowed, but that the implementation should default to (machine interpretable) way 2, i.e. try to download the linked file directly, and only revert to way 1 (human interaction) in case way 2 fails.

Reply:
> I prefer that either specification is allowed, but that the implementation should default to (machine interpretable) way 2, i.e. try to download the linked file directly, and only revert to way 1 (human interaction) in case way 2 fails.

+1 to this. it would be good to consider possible implementations, but the specification should not mandate one i think. as a specification here, the goal is to say there is a way to get to file X or contents of directory Y.